From b770823351c84c96a6f9be1cb9f7a8f6252e8423 Mon Sep 17 00:00:00 2001
From: David Daney <david.daney@cavium.com>
Date: Sat, 11 May 2019 19:54:05 +0200
Subject: [PATCH 201/345] MIPS: OCTEON:  Add sysfs hooks to add and remove
 CPUs.

Once a CPU is offline, we can do:

  echo cpu_num > /sys/devices/system/cpu/octeon_unplug

This makes the cpu available for oct-app-ctl or other uses. Doing:

  echo cpu_num > /sys/devices/system/cpu/octeon_plug

Restores it for use by the Linux kernel.

Signed-off-by: David Daney <david.daney@cavium.com>
Signed-off-by: Carlos Munoz <cmunoz@caviumnetworks.com>
---
 arch/mips/cavium-octeon/Makefile        |   8 +
 arch/mips/cavium-octeon/unplug-cpu-ll.S |  46 +++++
 arch/mips/cavium-octeon/unplug-cpu.c    | 231 ++++++++++++++++++++++++
 arch/mips/kernel/topology.c             |   2 +-
 drivers/base/cpu.c                      |   7 +
 5 files changed, 293 insertions(+), 1 deletion(-)
 create mode 100644 arch/mips/cavium-octeon/unplug-cpu-ll.S
 create mode 100644 arch/mips/cavium-octeon/unplug-cpu.c

diff --git a/arch/mips/cavium-octeon/Makefile b/arch/mips/cavium-octeon/Makefile
index d29207f332a8..8d4ab7d44368 100644
--- a/arch/mips/cavium-octeon/Makefile
+++ b/arch/mips/cavium-octeon/Makefile
@@ -28,3 +28,11 @@ obj-$(CONFIG_OCTEON_ERROR_INJECTOR)	+= octeon-error-injector.o
 obj-$(CONFIG_CAVIUM_OCTEON_ERROR_TREE)	+= octeon-error-tree.o octeon-78xx-errors.o \
 	octeon-73xx-errors.o octeon-7xxx-errors.o octeon-75xx-errors.o
 obj-$(CONFIG_SYSFS)                     += cacheinfo.o
+
+ifdef CONFIG_SMP
+ifdef CONFIG_SYSFS
+ifdef CONFIG_HOTPLUG_CPU
+obj-y += unplug-cpu.o unplug-cpu-ll.o
+endif
+endif
+endif
diff --git a/arch/mips/cavium-octeon/unplug-cpu-ll.S b/arch/mips/cavium-octeon/unplug-cpu-ll.S
new file mode 100644
index 000000000000..e48b75b0d282
--- /dev/null
+++ b/arch/mips/cavium-octeon/unplug-cpu-ll.S
@@ -0,0 +1,46 @@
+/*
+ * This file is subject to the terms and conditions of the GNU General Public
+ * License.  See the file "COPYING" in the main directory of this archive
+ * for more details.
+ *
+ * Copyright (C) 2014 Cavium, Inc.
+ */
+#include <asm/asm.h>
+#include <asm/regdef.h>
+#include <asm/mipsregs.h>
+
+#define	CVMCTL $9, 7
+
+	.p2align 3
+	.globl	octeon_replug_ll
+	.type	octeon_replug_ll, @function
+	.ent	octeon_replug_ll, 0
+	.set	noreorder
+	.set	noat
+octeon_replug_ll:
+	dmfc0	v0, CVMCTL
+	nop
+
+	ori	v0, 2
+	nop
+
+#ifdef CONFIG_CPU_BIG_ENDIAN
+	xori	v0, 2
+	nop
+#endif
+	dmtc0	v0, CVMCTL	/* Set proper endian mode for this kernel */
+	nop
+
+	/* Set a known operating mode */
+	mfc0	v0, CP0_STATUS
+	ori	v0, (ST0_IE | ST0_EXL | ST0_ERL | ST0_KX | ST0_SX | ST0_UX)
+	xori	v0, ST0_IE
+	mtc0	v0, CP0_STATUS
+	/* Wait for NMI, which will vector us to the startup code. */
+loop_wait:
+	wait
+	b	loop_wait
+	 nop
+
+	.end	octeon_replug_ll
+	.size	octeon_replug_ll, .-octeon_replug_ll
diff --git a/arch/mips/cavium-octeon/unplug-cpu.c b/arch/mips/cavium-octeon/unplug-cpu.c
new file mode 100644
index 000000000000..669bb0101066
--- /dev/null
+++ b/arch/mips/cavium-octeon/unplug-cpu.c
@@ -0,0 +1,231 @@
+/*
+ * This file is subject to the terms and conditions of the GNU General Public
+ * License.  See the file "COPYING" in the main directory of this archive
+ * for more details.
+ *
+ * Copyright (C) 2014 Cavium, Inc.
+ */
+#include <linux/module.h>
+#include <linux/init.h>
+#include <linux/cpu.h>
+#include <linux/device.h>
+#include <linux/percpu.h>
+
+#include <asm/octeon/octeon.h>
+#include <asm/octeon/cvmx-boot-vector.h>
+#include <asm/octeon/octeon-boot-info.h>
+#include <asm/octeon/cvmx-app-hotplug.h>
+#include <asm/octeon/cvmx-spinlock.h>
+#include <asm/octeon/cvmx-coremask.h>
+
+static struct cvmx_boot_vector_element *octeon_bootvector;
+static void *octeon_replug_ll_raw;
+asmlinkage void octeon_replug_ll(void);
+
+static struct cvmx_app_hotplug_global *hgp;
+static const cvmx_bootmem_named_block_desc_t *ccbi_desc;
+
+DECLARE_PER_CPU(struct cpu, cpu_devices);
+
+/* Need __ref to be able to call register_cpu().  This is OK as this
+ * file is only compiled for HOTPLUG_CPU so the resulting call to a
+ * __cpuinit function will always be valid.
+ */
+static ssize_t __ref plug_cpu_store(struct device *dev,
+				    struct device_attribute *attr,
+				    const char *buf, size_t count)
+{
+	int cpu, r, coreid, node;
+	unsigned long flags;
+	bool made_present = false;
+	bool is_available = false;
+
+	r = sscanf(buf, "%d", &cpu);
+
+	if (r != 1 || cpu < 0 || cpu >= NR_CPUS)
+		return -EINVAL;
+
+
+	cpu_maps_update_begin();
+
+	if (!cpu_present(cpu) && cpu_possible(cpu)) {
+		coreid = cpu_logical_map(cpu);
+
+		local_irq_save(flags);
+		cvmx_spinlock_lock(&hgp->hotplug_global_lock);
+		if (cvmx_coremask_is_core_set(&hgp->avail_coremask, coreid)) {
+			is_available = true;
+			cvmx_coremask_clear_core(&hgp->avail_coremask, coreid);
+		}
+		cvmx_spinlock_unlock(&hgp->hotplug_global_lock);
+		local_irq_restore(flags);
+		if (!is_available) {
+			pr_notice("CPU %d is not available for plugging\n", cpu);
+			goto not_available_out;
+		}
+
+		octeon_bootvector[coreid].target_ptr = (uint64_t)octeon_replug_ll_raw;
+		mb();
+		node = cvmx_coremask_core_to_node(coreid);
+		coreid = cvmx_coremask_core_on_node(coreid);
+		if (octeon_has_feature(OCTEON_FEATURE_CIU3))
+			cvmx_write_csr_node(node, CVMX_CIU3_NMI, (1ull << coreid));
+		else
+			cvmx_write_csr(CVMX_CIU_NMI, (1 << coreid));
+
+		set_cpu_present(cpu, true);
+		made_present = true;
+		pr_info("CPU %d now present\n", cpu);
+	}
+not_available_out:
+	cpu_maps_update_done();
+
+	if (made_present) {
+		struct cpu *c = &per_cpu(cpu_devices, cpu);
+		memset(c, 0, sizeof(struct cpu));
+		c->hotpluggable = 1;
+		r = register_cpu(c, cpu);
+		if (r)
+			pr_warn("unplug_cpu: register_cpu %d failed (%d)\n.", cpu, r);
+	}
+
+	return count;
+}
+
+static ssize_t unplug_cpu_store(struct device *dev,
+				struct device_attribute *attr,
+				const char *buf, size_t count)
+{
+	int cpu, r;
+	bool made_not_present = false;
+	unsigned long flags;
+
+	r = sscanf(buf, "%d", &cpu);
+
+	if (r != 1 || cpu < 0 || cpu >= NR_CPUS)
+		return -EINVAL;
+
+	cpu_maps_update_begin();
+
+	if (!cpu_online(cpu) && cpu_present(cpu)) {
+		pr_info("CPU %d now not present\n", cpu);
+		set_cpu_present(cpu, false);
+		made_not_present = true;
+	}
+
+	cpu_maps_update_done();
+
+	if (made_not_present) {
+		int coreid = cpu_logical_map(cpu);
+		struct cpu *c = &per_cpu(cpu_devices, cpu);
+		unregister_cpu(c);
+
+		local_irq_save(flags);
+		cvmx_spinlock_lock(&hgp->hotplug_global_lock);
+		cvmx_coremask_set_core(&hgp->avail_coremask, coreid);
+		cvmx_spinlock_unlock(&hgp->hotplug_global_lock);
+		local_irq_restore(flags);
+	}
+
+	return count;
+}
+
+static ssize_t unplug_cpu_print(struct device *dev,
+				struct device_attribute *attr, char *buf)
+{
+	return sprintf(buf, "hello\n");
+}
+
+DEVICE_ATTR(octeon_plug, 0644, unplug_cpu_print, plug_cpu_store);
+DEVICE_ATTR(octeon_unplug, 0644, unplug_cpu_print, unplug_cpu_store);
+
+/* the following function will work ONLY with size%8 = 0 */
+static
+int __cvmx_copy_from_bootmem(int64_t bootmem_src_addr, void *dst_ptr, int size)
+{
+	int i;
+	int64_t base_addr = (1ull << 63) | bootmem_src_addr;
+	int64_t *ptr64 = dst_ptr;
+
+	for (i = 0; i < size/8; i++) {
+		ptr64[i] = cvmx_read64_int64(base_addr);
+		base_addr += 8;
+	}
+	return 0;
+}
+
+static void __init octeon_hotplug_global_init(void *arg)
+{
+	struct linux_app_boot_info *labi;
+	cvmx_app_hotplug_global_t *hgp = arg;
+	cvmx_cores_common_bootinfo_t ccbi;
+
+	memset(hgp, 0, CVMX_APP_HOTPLUG_INFO_REGION_SIZE);
+
+	hgp->magic_version = CVMX_HOTPLUG_MAGIC_VERSION;
+
+	cvmx_spinlock_init(&hgp->hotplug_global_lock);
+
+	/* Get legacy LABI data structure for initial parameters */
+	labi = phys_to_virt(LABI_ADDR_IN_BOOTLOADER);
+
+	/* Initialize available cores from LABI is limited to 32 cores
+	 * - try to do not use it - instead do ... */
+	if (ccbi_desc) { /* 'common bootinfo' named block is found - use it*/
+		__cvmx_copy_from_bootmem(ccbi_desc->base_addr, &ccbi,
+					sizeof(cvmx_cores_common_bootinfo_t));
+		/* Validate signature */
+		if (ccbi.magic != CVMX_COMMON_BOOTINFO_MAGIC)
+			return; /* if 'magic' does not match - exit */
+		/* the members from the initial(1) version are always valid */
+		/* only hgp->avail_coremask is need - fill it in */
+		cvmx_coremask_copy(&hgp->avail_coremask, &ccbi.avail_coremask);
+		/* the extra (version) members (if any) are valid when
+		 * (2 <= ccbi.version <= CVMX_COMMON_BOOTINFO_VERSION)
+		 * if (ccbi.version >= 2) { xxx = ccbi.ver2_member; }
+		 */
+	} else { /* the older bootloaders provide only labi->avail_coremask */
+		/* Validate signature */
+		if (labi->labi_signature != LABI_SIGNATURE)
+			return;
+		cvmx_coremask_set64(&hgp->avail_coremask,
+					(uint64_t) labi->avail_coremask);
+	}
+}
+
+static int __init unplug_cpu_init(void)
+{
+	unsigned long t;
+
+	octeon_bootvector = cvmx_boot_vector_get();
+	if (!octeon_bootvector) {
+		pr_err("Error: Cannot allocate boot vector.\n");
+		return -ENOMEM;
+	}
+	t = __pa_symbol(octeon_replug_ll);
+	octeon_replug_ll_raw = phys_to_virt(t);
+
+	/* the 'common bootinfo' named block should be found/copied before
+	 * creating hotplug named block,
+	 * because avail_coremask is copied from it
+	 */
+	ccbi_desc =
+		cvmx_bootmem_find_named_block(CVMX_APP_COMMON_BOOTINFO_NAME);
+	if (!ccbi_desc)
+		pr_info("Info: cvmx_bootmem_find_named_block(%s) not found.\n",
+						CVMX_APP_COMMON_BOOTINFO_NAME);
+
+	hgp = cvmx_bootmem_alloc_named_range_once(
+		CVMX_APP_HOTPLUG_INFO_REGION_SIZE,
+		0x0, 1ull << 29, 0,
+		CVMX_APP_HOTPLUG_INFO_REGION_NAME,
+		octeon_hotplug_global_init);
+
+	if (!hgp) {
+		pr_err("Error: cvmx_bootmem_alloc_named_range_once(%s)\n",
+		       CVMX_APP_HOTPLUG_INFO_REGION_NAME);
+		return -ENOMEM;
+	}
+	return 0;
+}
+module_init(unplug_cpu_init);
diff --git a/arch/mips/kernel/topology.c b/arch/mips/kernel/topology.c
index cd3e1f82e1a5..9c81772a5f99 100644
--- a/arch/mips/kernel/topology.c
+++ b/arch/mips/kernel/topology.c
@@ -6,7 +6,7 @@
 #include <linux/nodemask.h>
 #include <linux/percpu.h>
 
-static DEFINE_PER_CPU(struct cpu, cpu_devices);
+DEFINE_PER_CPU(struct cpu, cpu_devices);
 
 static int __init topology_init(void)
 {
diff --git a/drivers/base/cpu.c b/drivers/base/cpu.c
index 93758b528d8f..4e447f0b4312 100644
--- a/drivers/base/cpu.c
+++ b/drivers/base/cpu.c
@@ -449,6 +449,9 @@ EXPORT_SYMBOL_GPL(cpu_device_create);
 static DEVICE_ATTR(modalias, 0444, print_cpu_modalias, NULL);
 #endif
 
+extern struct device_attribute dev_attr_octeon_plug;
+extern struct device_attribute dev_attr_octeon_unplug;
+
 static struct attribute *cpu_root_attrs[] = {
 #ifdef CONFIG_ARCH_CPU_PROBE_RELEASE
 	&dev_attr_probe.attr,
@@ -465,6 +468,10 @@ static struct attribute *cpu_root_attrs[] = {
 #endif
 #ifdef CONFIG_GENERIC_CPU_AUTOPROBE
 	&dev_attr_modalias.attr,
+#endif
+#if defined(CONFIG_CAVIUM_OCTEON_SOC) && defined(CONFIG_HOTPLUG_CPU)
+	&dev_attr_octeon_plug.attr,
+	&dev_attr_octeon_unplug.attr,
 #endif
 	NULL
 };
-- 
2.25.1

